### ISSUE TITLE
Range algorithms don't work when used with unwrappable iterators and custom sentinels

### ISSUE THREAD

**Issue by JMazurkiewicz** [microsoft/STL#3010]
## Description

Range algorithms, constructors (#2806) and iterator operations won't compile when used with unwrappable iterator "`It`" (e.g. `string::const_iterator`) and custom sentinel "`Se`", such that `sentinel_for<Se, X>` is true **if and only if** `X` is convertible to `It`.

### Custom sentinels and unwrappable iterators don't work together

```c++
#include <algorithm>
#include <iostream>
#include <iterator>
#include <string>

using namespace std;

struct question_sentinel_t {
    constexpr bool operator==(string::const_iterator it) const noexcept {
        return *it == '?' || *it == '\0';
    }
};

int main() {
    string str = "qwerty?uiopasdfghjklzxcvbnm";
    ranges::copy(str.begin(), question_sentinel_t{}, ostreambuf_iterator<char>{cout});
    cout << '\n';
}
```

* **Expected behavior:** This code should compile ([compiler explorer](https://godbolt.org/z/EYrrbr1EY)).
* **Expected output:** `qwerty`

### Ranges with custom sentinels and unwrappable iterators cannot be used as inputs to algorithms

```c++
#include <algorithm>
#include <iostream>
#include <ranges>
#include <string>

using namespace std;

template<ranges::view V>
    requires ranges::input_range<V>
class question_view : public ranges::view_interface<question_view<V>> {
private:
    struct sentinel {
        constexpr bool operator==(const ranges::iterator_t<V>& it) const noexcept {
            return *it == '?' || *it == '\0';
        }
    };

public:
    constexpr explicit question_view(V base)
        : m_base(std::move(base)) { }
    
    constexpr auto begin() const {
        return ranges::begin(m_base);
    }

    constexpr auto end() const noexcept {
        return sentinel{};
    }

private:
    V m_base;
};

template<typename R>
question_view(R&&) -> question_view<views::all_t<R>>;

int main() {
    string str = "qwerty?uiopasdfghjklzxcvbnm";
    auto v = question_view{str};
    static_assert(ranges::view<decltype(v)>);
    ranges::copy(v, ostreambuf_iterator<char>{cout});
    cout << '\n';
}
```

* **Expected behavior:** This code should compile ([compiler explorer](https://godbolt.org/z/K6n16xcx3)).
* **Expected output:** `qwerty`

### Conversions from ranges to containers do not work

```c++
#include <algorithm>
#include <forward_list>
#include <iostream>
#include <ranges>
#include <string>

using namespace std;

template<ranges::view V>
    requires ranges::input_range<V>
class question_view : public ranges::view_interface<question_view<V>> {
private:
    struct sentinel {
        constexpr bool operator==(const ranges::iterator_t<V>& it) const noexcept {
            return *it == '?' || *it == '\0';
        }
    };

public:
    constexpr explicit question_view(V base)
        : m_base(std::move(base)) { }
    
    constexpr auto begin() const {
        return ranges::begin(m_base);
    }

    constexpr auto end() const noexcept {
        return sentinel{};
    }

private:
    V m_base;
};

template<typename R>
question_view(R&&) -> question_view<views::all_t<R>>;

int main() {
    auto str = "qwerty?uiopasdfghjklzxcvbnm"s;
    (void) (question_view{str} | ranges::to<forward_list>());
}
```

**Expected behavior:** This code should compile (~~assuming #2806 was merged~~).

## Summary

1. Some algorithms (possibly all of them) don't work with custom sentinels and unwrappable iterators. I've tested:
    * `ranges::all_of`, `ranges::any_of`, `ranges::none_of`,
    * `ranges::copy`,
    * `ranges::sort`.
3. Some functions from `<iterator>` header don't work too (e.g. `ranges::distance`),
4. Same thing applies to range constructors (and possibly other operations) from #2806.

**Comment by frederick-vs-ja** [comment]
I think this has been reported as #2591.

**Comment by CaseyCarter** [comment]
> I think this has been reported as #2591.

Yes, although it's far from obvious: this is a good description of the general issue, whereas #2591 is a particular occurrence. I'll keep this around for now and link it up to the meta-issue.

**Comment by strega-nil-ms** [comment]
Working on this in #3024

**Comment by JMazurkiewicz** [comment]
I think that this and #2591 are not bugs anymore (thanks to #3024).

**Comment by strega-nil-ms** [comment]
Fixed by #3024 (you're right, thanks @Jmazurkiewicz!)
